src/runtime_src/xdp/profile/plugin/power/power_plugin.cpp

```diff
@@ -0,0 +1,141 @@
+/**
+ * Copyright (C) 2020 Xilinx, Inc
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"). You may
+ * not use this file except in compliance with the License. A copy of the
+ * License is located at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+#define XDP_SOURCE
+
+#include "xdp/profile/plugin/power/power_plugin.h"
+#include "xdp/profile/writer/power/power_writer.h"
+#include "core/common/system.h"
+#include "core/common/time.h"
+#include "core/include/experimental/xrt-next.h"
+
+namespace xdp {
+
+  const char* PowerProfilingPlugin::powerFiles[] = 
+    {
+      "xmc_12v_aux_curr",
+      "xmc_12v_aux_vol",
+      "xmc_12v_pex_curr",
+      "xmc_12v_pex_vol",
+      "xmc_vccint_curr",
+      "xmc_vccint_vol",
+      "xmc_3v3_pex_curr",
+      "xmc_3v3_pex_vol",
+      "xmc_cage_temp0",
+      "xmc_cage_temp1",
+      "xmc_cage_temp2",
+      "xmc_cage_temp3",
+      "xmc_dimm_temp0",
+      "xmc_dimm_temp1",
+      "xmc_dimm_temp2",
+      "xmc_dimm_temp3",
+      "xmc_fan_temp",
+      "xmc_fpga_temp",
+      "xmc_hbm_temp",
+      "xmc_se98_temp0",
+      "xmc_se98_temp1",
+      "xmc_se98_temp2",
+      "xmc_vccint_temp",
+      "xmc_fan_rpm"      
+    } ;
+
+  PowerProfilingPlugin::PowerProfilingPlugin() : XDPPlugin(), keepPolling(true)
+  {
+    db->registerPlugin(this) ;
+   
+    // Just like HAL level device profiling, go through the devices 
+    //  that exist in order to find all of the sysfs paths
+    uint64_t index = 0 ;
+    void* handle = xclOpen(index, "/dev/null", XCL_INFO) ;
+    while (handle != nullptr)
+    {
+      // For each device, keep track of the paths to the sysfs files
+      std::vector<std::string> paths ;
+      for (auto f : powerFiles)
+      {
+	char sysfsPath[512] ;
+	xclGetSysfsPath(handle, "xmc", f, sysfsPath, 512) ;
```

Comment:
need a check here in case xclGetSysfsPath returns a null path. In that case we need to skip the device

Reply:
In order to make sure all devices have the same number of entries (one per file defined in powerFiles) and all the data is aligned, I want to push the null path or invalid paths into our vector.  Later, when we poll the data, if the file cannot be opened we record the value as 0.